File: cpp/llama.cpp
#include <iostream>
#include <cmath>

#include <torch/torch.h>
#include <sentencepiece_processor.h>

int sample(const torch::Tensor& logits, double temperature) {
  if (temperature < 1e-6) {
    return logits.argmax().item<int>();
  } else {
    torch::Tensor probs = torch::softmax(logits / temperature, -1);
    return torch::multinomial(probs, 1).item<int>();
  }
}

torch::Tensor precompute_freqs_cis(int dim, int end, float theta = 10000.0) {
  torch::Tensor indices = torch::arange(0, dim, 2);
  indices.slice(0, dim/2).to(torch::kFloat);
  torch::Tensor freq_denominator = torch::pow(theta, indices / dim);
  torch::Tensor freqs = 1.0 / freq_denominator;
  torch::Tensor t = torch::arange(end, torch::TensorOptions().device(freqs.device())); // type: ignore
  freqs = torch::outer(t, freqs).to(torch::kFloat32); // type: ignore
  torch::Tensor freqs_cis = torch::polar(torch::ones_like(freqs), freqs);
  return freqs_cis;
}

torch::Tensor reshape_for_broadcast(const torch::Tensor& freqs_cis, const torch::Tensor& x) {
    int ndim = x.dim();
    assert(0 <= 1 && 1 < ndim);
    assert(freqs_cis.sizes() == torch::IntArrayRef({x.size(1), x.size(-1)}));

    std::vector<int64_t> shape;
    for (int i = 0; i < ndim; ++i) {
        shape.push_back((i == 1 || i == ndim - 1) ? x.size(i) : 1);
    }

    return freqs_cis.view(shape);
}

std::tuple<torch::Tensor, torch::Tensor> apply_rotary_emb(torch::Tensor xq, torch::Tensor xk, torch::Tensor freqs_cis) {
    // Convert xq and xk to complex numbers
    torch::Tensor xq_ = torch::view_as_complex(xq.to(torch::kFloat32).reshape({xq.size(0), xq.size(1), xq.size(2), -1, 2}));
    torch::Tensor xk_ = torch::view_as_complex(xk.to(torch::kFloat32).reshape({xk.size(0), xk.size(1), xk.size(2), -1, 2}));

    // Reshape freqs_cis for broadcasting
    freqs_cis = reshape_for_broadcast(freqs_cis, xq_);

    // Apply rotary embeddings
    torch::Tensor xq_out = torch::view_as_real(xq_ * freqs_cis).flatten(3);
    torch::Tensor xk_out = torch::view_as_real(xk_ * freqs_cis).flatten(3);

    // Convert back to the original data type and return
    return std::make_tuple(xq_out.to(xq.scalar_type()), xk_out.to(xk.scalar_type()));
}

class RMSNormImpl : public torch::nn::Module {
public:
  RMSNormImpl(int64_t dim, double eps = 1e-6)
    : eps(eps),
      weight(torch::ones({dim}, torch::TensorOptions().device(torch::kCUDA))),
      dim(dim) {
    register_parameter("weight", weight);
  }

  torch::Tensor forward(torch::Tensor x) {
    torch::Tensor output = _norm(x.to(torch::kFloat32)).to(x.scalar_type());
    return output * weight;
  }
private:
  double eps;
  torch::Tensor weight;
  int64_t dim;
  torch::Tensor _norm(torch::Tensor x) {
    return x * torch::rsqrt(x.pow(2).mean(-1, true) + eps);
  }
};
TORCH_MODULE(RMSNorm);

class AttentionImpl : public torch::nn::Module {
public:
  AttentionImpl(int64_t dim, int64_t n_heads)
    : wq(torch::nn::Linear(dim, dim)),
      wk(torch::nn::Linear(dim, dim)),
      wv(torch::nn::Linear(dim, dim)),
      wo(torch::nn::Linear(dim, dim)),
      n_heads(n_heads),
      head_dim(dim / n_heads) {
    register_module("wq", wq);
    register_module("wk", wk);
    register_module("wv", wv);
    register_module("wo", wo);
  }

  torch::Tensor forward(torch::Tensor x, int64_t start_pos, torch::Tensor freqs_cis, torch::Tensor mask) {
    int64_t bsz = x.size(0);
    int64_t seqlen = x.size(1);

    torch::Tensor xq = wq(x);
    torch::Tensor xk = wk(x);
    torch::Tensor xv = wv(x);

    xq = xq.view({bsz, seqlen, n_heads, head_dim});
    xk = xk.view({bsz, seqlen, n_heads, head_dim});
    xv = xv.view({bsz, seqlen, n_heads, head_dim});

    auto result = apply_rotary_emb(xq, xk, freqs_cis);
    xq = std::get<0>(result);
    xk = std::get<1>(result);

    xq = xq.transpose(1, 2);
    torch::Tensor keys = xk.transpose(1, 2);
    torch::Tensor values = xv.transpose(1, 2);

    torch::Tensor scores = torch::matmul(xq, keys.transpose(2, 3)) / std::sqrt(static_cast<float>(head_dim));

    if (mask.defined()) {
      scores += mask;
    }

    scores = torch::softmax(scores, -1);
    torch::Tensor output = torch::matmul(scores, values).transpose(1, 2).contiguous().view({bsz, seqlen, -1});
    return wo(output);
  }
private:
  torch::nn::Linear wq, wk, wv, wo;
  int64_t dim, n_heads, head_dim;
};
TORCH_MODULE(Attention);

class FeedForwardImpl : public torch::nn::Module {
public:
  FeedForwardImpl(int64_t dim, int64_t hidden_dim, int64_t multiple_of)
    : dim(dim) {
    hidden_dim = multiple_of * ((hidden_dim + multiple_of - 1) / multiple_of);
    w1 = register_module("w1", torch::nn::Linear(dim, hidden_dim));
    w2 = register_module("w2", torch::nn::Linear(hidden_dim, dim));
    w3 = register_module("w3", torch::nn::Linear(dim, hidden_dim));
  }

  torch::Tensor forward(torch::Tensor x) {
    return w2(torch::silu(w1(x)) * w3(x));
  }

private:
  torch::nn::Linear w1{nullptr}, w2{nullptr}, w3{nullptr};
  int64_t dim, hidden_dim;
};
TORCH_MODULE(FeedForward);

class TransformerBlockImpl : public torch::nn::Module {
public:
  TransformerBlockImpl(int64_t dim, int64_t multiple_of, int64_t n_heads, double norm_eps)
    : attention(Attention(dim, n_heads)),
      feed_forward(FeedForward(dim, 4 * dim, multiple_of)),
      attention_norm(RMSNorm(dim, norm_eps)),
      ffn_norm(RMSNorm(dim, norm_eps)) {
    register_module("attention", attention);
    register_module("feed_forward", feed_forward);
    register_module("attention_norm", attention_norm);
    register_module("ffn_norm", ffn_norm);
  }

  torch::Tensor forward(torch::Tensor x, int64_t start_pos, torch::Tensor freqs_cis, torch::Tensor mask) {
    torch::Tensor h = x + attention->forward(attention_norm->forward(x), start_pos, freqs_cis, mask);
    return h + feed_forward->forward(ffn_norm->forward(h));
  }
private:
  Attention attention;
  FeedForward feed_forward;
  RMSNorm attention_norm, ffn_norm;
};
TORCH_MODULE(TransformerBlock);

class TransformerImpl : public torch::nn::Module {
public:
  TransformerImpl(int64_t dim, int64_t multiple_of, int64_t n_heads, int64_t n_layers, 
    double norm_eps, int64_t vocab_size, int64_t max_batch_size=32, int64_t max_seq_len=2048) 
    : dim(dim),
      multiple_of(multiple_of),
      n_heads(n_heads),
      n_layers(n_layers),
      norm_eps(norm_eps),
      max_batch_size(max_batch_size),
      max_seq_len(max_seq_len) {
    // for (int i = 0; i < n_layers; ++i) {
    //   layers.push_back(TransformerBlock(dim, multiple_of, n_heads, norm_eps));
    // }
    norm = register_module("norm", RMSNorm(dim, norm_eps));
    tok_embeddings = register_module("tok_embeddings", torch::nn::Embedding(vocab_size, dim));
    layer = register_module("block", TransformerBlock(dim, multiple_of, n_heads, norm_eps));
    output = register_module("output", torch::nn::Linear(dim, vocab_size));
    freqs_cis = precompute_freqs_cis(dim / n_heads, max_seq_len);
  }

  torch::Tensor forward(torch::Tensor tokens, int64_t start_pos, torch::Device device=torch::kCPU) {
    int64_t _bsz = tokens.sizes()[0];
    int64_t seqlen = tokens.sizes()[1];
    
    auto h = tok_embeddings->forward(tokens);
    auto freqs_cis_sliced = freqs_cis.slice(0, start_pos, start_pos + seqlen).to(device);

    torch::Tensor mask = torch::empty({1, 1, seqlen, seqlen}, device).fill_(std::numeric_limits<float>::lowest());
    mask.triu_(start_pos + 1);

    // for (int layer = 0; layer < n_layers; ++layer) {
    //   h = layers[layer]->forward(h, start_pos, freqs_cis_sliced, mask);
    // }
    h = layer->forward(h, start_pos, freqs_cis_sliced, mask);
    return output->forward((norm->forward(h)).narrow(1, -1, 1).squeeze(1)); // only compute the last logits
  }
private:
  int64_t dim, multiple_of, n_heads, n_layers, cache_size, max_batch_size, max_seq_len;
  double norm_eps;
  std::string param;
  // std::vector<TransformerBlock> layers;
  TransformerBlock layer{nullptr};
  RMSNorm norm{nullptr};
  torch::nn::Embedding tok_embeddings{nullptr};
  torch::nn::Linear output{nullptr};
  torch::Tensor freqs_cis;
};
TORCH_MODULE(Transformer);

int main() {
  torch::manual_seed(1337);
  const std::string PARAM = "7B";
  const int VOCAB_SIZE = 32000;
  torch::Device device = torch::kCPU;
  if (torch::cuda::is_available()) {
    std::cout << "Device: CUDA" << std::endl;
    device = torch::kCUDA;
  }
  sentencepiece::SentencePieceProcessor processor;
  const auto status = processor.Load("../../weights/tokenizer.model");
  if (!status.ok()) {
    std::cerr << status.ToString() << std::endl;
  }

  Transformer model(512, 256, 8, 8, 1e-6, VOCAB_SIZE);
  model->to(device);
  // load state dict
  std::string prompt = "Elon Musk is ";

  std::vector<int> ids;
  processor.Encode(prompt, &ids);

  while (true) {
    torch::InferenceMode guard;
    torch::Tensor logits;
    torch::Tensor toks = torch::from_blob(ids.data(), {static_cast<int64_t>(ids.size())}, torch::kInt32);
    {
      torch::NoGradGuard no_grad;
      torch::Tensor input_tensor = toks.unsqueeze(0).to(device);
      logits = model(input_tensor, 0, device);
    }
    int tok = sample(logits, 0.7);
    ids.push_back(tok);

    std::string text;
    processor.Decode(ids, &text);
    std::cout << text << std::endl;
  }

  return 0;
}

File: llama.py
import os
import math
from pathlib import Path

import numpy as np
import torch
torch.manual_seed(1337)
import torch.nn as nn
from torch.nn import functional as F

from utils import Timing

def sample(logits, temperature):
    if temperature < 1e-6:
        return int(logits.item().argmax())
    else:
        probs = F.softmax(logits / temperature, dim=-1)
        return torch.multinomial(probs, num_samples=1).item()

def precompute_freqs_cis(dim: int, end: int, theta: float = 10000.0):
    freqs = 1.0 / (theta ** (torch.arange(0, dim, 2)[: (dim // 2)].float() / dim))
    t = torch.arange(end, device=freqs.device)  # type: ignore
    freqs = torch.outer(t, freqs).float()  # type: ignore
    freqs_cis = torch.polar(torch.ones_like(freqs), freqs)  # complex64
    return freqs_cis

def reshape_for_broadcast(freqs_cis: torch.Tensor, x: torch.Tensor):
    ndim = x.ndim
    assert 0 <= 1 < ndim
    assert freqs_cis.shape == (x.shape[1], x.shape[-1])
    shape = [d if i == 1 or i == ndim - 1 else 1 for i, d in enumerate(x.shape)]
    return freqs_cis.view(*shape)

def apply_rotary_emb(xq, xk, freqs_cis, device=None):
    xq_ = torch.view_as_complex(xq.float().reshape(*xq.shape[:-1], -1, 2))
    xk_ = torch.view_as_complex(xk.float().reshape(*xk.shape[:-1], -1, 2))
    freqs_cis = reshape_for_broadcast(freqs_cis, xq_)
    xq_out = torch.view_as_real(xq_ * freqs_cis).flatten(3)
    xk_out = torch.view_as_real(xk_ * freqs_cis).flatten(3)
    return xq_out.type_as(xq), xk_out.type_as(xk)

class RMSNorm(nn.Module):
    def __init__(self, dim: int, eps: float = 1e-6):
        super().__init__()
        self.eps = eps
        self.weight = nn.Parameter(torch.ones(dim))
        self.dim = dim

    def _norm(self, x):
        return x * torch.rsqrt(x.pow(2).mean(-1, keepdim=True) + self.eps)
    
    def forward(self, x):
        output = self._norm(x.float()).type_as(x)
        return output * self.weight
    
    def load_weight(self, data, offset, device=None):
        self.weight.data = data[offset : offset+self.dim]
        # return offset + self.dim

class Attention(nn.Module):
    def __init__(self, dim, n_heads, n_layers):
        super().__init__()
        self.wq, self.wk, self.wv, self.wo = [nn.Linear(dim, dim, bias=False) for _ in range(4)]
        self.buf = nn.Parameter(torch.ones(dim*dim//2))
        self.dim = dim
        self.n_heads = n_heads
        self.head_dim = dim // n_heads
        self.cache_k = [None for _ in range(n_layers)]
        self.cache_v = [None for _ in range(n_layers)]
        
    def forward(self, x, start_pos, freqs_cis, mask, layer):
        bsz, seqlen, _ = x.shape
        xq, xk, xv = self.wq(x), self.wk(x), self.wv(x)
        xq, xk, xv = [x.view(bsz, seqlen, self.n_heads, self.head_dim) for x in (xq, xk, xv)]
        xq, xk = apply_rotary_emb(xq, xk, freqs_cis=freqs_cis)

        # TODO: kv caching is broken
        if start_pos == 0:
            keys, values = xk, xv
        else:
            assert start_pos == self.cache_k[layer].shape[1] and start_pos == self.cache_v[layer].shape[1], f'{start_pos} {self.cache_k[layer].shape}'
            keys, values = torch.cat((self.cache_k[layer], xk), dim=1), torch.cat((self.cache_v[layer], xv), dim=1)
        
        self.cache_k[layer] = keys
        self.cache_v[layer] = values

        xq = xq.transpose(1, 2)
        keys = keys.transpose(1, 2)
        values = values.transpose(1, 2)
        scores = torch.matmul(xq, keys.transpose(2, 3)) / math.sqrt(self.head_dim)
        if mask is not None:
            scores += mask
        scores = F.softmax(scores, dim=-1)
        output = torch.matmul(scores, values).transpose(1, 2).contiguous().view(bsz, seqlen, -1)

        return self.wo(output)
    
    def load_weight(self, data, offset, device='cuda'):
        n_param = self.dim * self.dim

        self.wq.weight.data = data[offset : offset+n_param].view(self.dim, self.dim)
        offset += n_param
        self.wk.weight.data = data[offset : offset+n_param].view(self.dim, self.dim)
        offset += n_param
        self.wv.weight.data = data[offset : offset+n_param].view(self.dim, self.dim)
        offset += n_param
        self.wo.weight.data = data[offset : offset+n_param].view(self.dim, self.dim)
    
class FeedForward(nn.Module):
    def __init__(self, dim, hidden_dim, multiple_of):
        super().__init__()
        hidden_dim = int(2 * hidden_dim / 3)
        hidden_dim = multiple_of * ((hidden_dim + multiple_of - 1) // multiple_of)
        self.w1 = nn.Linear(dim, hidden_dim, bias=False)
        self.w2 = nn.Linear(hidden_dim, dim, bias=False)
        self.w3 = nn.Linear(dim, hidden_dim, bias=False)

        self.dim = dim
        self.hidden_dim = hidden_dim

    def __call__(self, x):
        return self.w2(F.silu(self.w1(x)) * self.w3(x))
    
    def load_weight(self, data, offset, device=None):
        n_param = self.dim * self.hidden_dim
        self.w1.weight.data = data[offset : offset+n_param].view(self.hidden_dim, self.dim)
        offset += n_param
        self.w2.weight.data = data[offset : offset+n_param].view(self.dim, self.hidden_dim)
        offset += n_param
        self.w3.weight.data = data[offset : offset+n_param].view(self.hidden_dim, self.dim)
    
class TransformerBlock(nn.Module):
    def __init__(self, dim, multiple_of, n_heads, norm_eps, param, n_layers):
        super().__init__()
        self.dim = dim
        self.attention = Attention(dim, n_heads, n_layers)
        self.feed_forward = FeedForward(dim, 4*dim, multiple_of)
        self.attention_norm = RMSNorm(dim, norm_eps)
        self.ffn_norm = RMSNorm(dim, norm_eps)
        self.serialized_dir = f'serialized/{param}'

    def forward(self, x, start_pos, freqs_cis, mask, layer):
        h = x + self.attention(self.attention_norm(x), start_pos, freqs_cis, mask, layer)
        return h + self.feed_forward(self.ffn_norm(h))
    
    def fetch(self, layer: int, device=None):
        data = np.memmap(os.path.join(self.serialized_dir, f'layer{layer}.bin'), dtype=np.float16, mode='r')
        offset = [0, 4*self.dim**2, 4*self.dim**2+3*self.dim*self.feed_forward.hidden_dim, 4*self.dim**2+3*self.dim*self.feed_forward.hidden_dim+self.dim]
        layer_weight = torch.tensor(data, dtype=torch.float16, device=device)

        self.attention.load_weight(layer_weight, offset[0])
        self.feed_forward.load_weight(layer_weight, offset[1])
        self.attention_norm.load_weight(layer_weight, offset[2])
        self.ffn_norm.load_weight(layer_weight, offset[3])

class Transformer(nn.Module):
    def __init__(self, dim, multiple_of, n_heads, n_layers, norm_eps, vocab_size, max_batch_size=32, max_seq_len=2048, param='7B'):
        super().__init__()
        self.n_layers = n_layers
        self.layer_cache = [TransformerBlock(dim, multiple_of, n_heads, norm_eps, param, n_layers)]
        self.norm = RMSNorm(dim, norm_eps)
        self.tok_embeddings = nn.Embedding(vocab_size, dim)
        self.output = nn.Linear(dim, vocab_size, bias=False)
        self.freqs_cis = precompute_freqs_cis(dim // n_heads, max_seq_len)

    def forward(self, tokens, start_pos, device=None):
        _bsz, seqlen = tokens.shape
        h = self.tok_embeddings(tokens)
        freqs_cis = self.freqs_cis[start_pos : start_pos + seqlen].to(device)

        mask = None
        if seqlen > 1:
            mask = torch.full((1, 1, seqlen, seqlen), float('-inf'), device=tokens.device)
            mask = torch.triu(mask, diagonal=start_pos+1)

        for layer in range(self.n_layers):
            # with Timing('IO: '):
            self.layer_cache[0].fetch(layer, device) # TODO: async execution
            # with Timing('Compute: '):
            h = self.layer_cache[0](h, start_pos, freqs_cis, mask, layer)

        return self.output(self.norm(h)[:, -1, :]) # only compute the last logits
    

# **** files and arguments ****

PARAM = '7B'
WEIGHTS_DIR = Path("weights")
TOKENIZER_FILENAME = WEIGHTS_DIR / "tokenizer.model"
VOCAB_SIZE = 32000

args_7B = {"dim": 4096, "multiple_of": 256, "n_heads": 32, "n_layers": 32, "norm_eps": 1e-06, "vocab_size": VOCAB_SIZE, "param": "7B"}
args_13B = {"dim": 5120, "multiple_of": 256, "n_heads": 40, "n_layers": 40, "norm_eps": 1e-06, "vocab_size": VOCAB_SIZE, "param": "13B"}
args_30B = {"dim": 6656, "multiple_of": 256, "n_heads": 52, "n_layers": 60, "norm_eps": 1e-06, "vocab_size": VOCAB_SIZE, "param": "30B"}
args_65B = {"dim": 8192, "multiple_of": 256, "n_heads": 64, "n_layers": 80, "norm_eps": 1e-06, "vocab_size": VOCAB_SIZE, "param": "65B"}

args = {
    '7B': args_7B,
    '13B': args_13B,
    '30B': args_30B,
    '65B': args_65B
}

if __name__ == '__main__':
    from sentencepiece import SentencePieceProcessor
    sp_model = SentencePieceProcessor(model_file=str(TOKENIZER_FILENAME))
    assert sp_model.vocab_size() == VOCAB_SIZE

    device = 'cuda' if torch.cuda.is_available else 'cpu'
    print(f'Device: {device}')
    
    model = Transformer(**args[PARAM]).half().to(device)
    model.load_state_dict(torch.load(f'serialized/{PARAM}/io.pt'))

    prompt = 'Elon Musk is '
    toks = [sp_model.bos_id()] + sp_model.encode(prompt)
    start_pos = 0

    while True:
        with torch.inference_mode(), Timing('== '):
            logits = model(torch.tensor(toks[start_pos:]).unsqueeze(dim=0).to(device), start_pos, device)
        tok = sample(logits, 0.7)
        start_pos = len(toks)
        toks.append(tok)
        cur = sp_model.decode(toks)
        print(cur)
